### PR TITLE
Custom service name

### LINKED ISSUE #27: Specify custom service name
I would like to be able to specify custom service name, instead of using marathon app id. Currently you need to add `consul:true` label to be visible to `marathon-consul`. We could change this syntax to: `consul:custom_service_name`. `true` could be a *protected keyword* for when you don't want any custom name (and to maintain backwards compatibility).

### PR DESCRIPTION
**Pull request by dankraw**
Fixes #27 

An important change has been introduced: `app_terminated_event` is no more being handled, `deployment_info` and `deployment_step_success` handling added instead.

### CODE REVIEW

File: consul/consul.go [hunk 1]
```diff
@@ -13,9 +12,9 @@ import (
 
 type ConsulServices interface {
 	GetAllServices() ([]*consulapi.CatalogService, error)
-	GetServices(name tasks.AppId) ([]*consulapi.CatalogService, error)
-	Register(task *tasks.Task, app *apps.App) error
-	Deregister(serviceId tasks.Id, agentAddress string) error
+	GetServices(name string) ([]*consulapi.CatalogService, error)
+	Register(task *apps.Task, app *apps.App) error
```
Comment (janisz) [inline consul/consul.go]: I think we can get rid of `task` parameter since it could be embedded in `app`

Comment (dankraw) [inline consul/consul.go]: this will end up with a big code refactor, I would suggest doing this in the future. :)

File: apps/app.go [hunk 2]
```diff
@@ -26,13 +26,36 @@ type AppsResponse struct {
...
 
 func (app *App) IsConsulApp() bool {
-	value, ok := app.Labels["consul"]
-	return ok && value == "true"
+	_, ok := app.Labels["consul"]
+	return ok
+}
+
+func (app *App) ConsulServiceName() string {
+	if value, ok := app.Labels["consul"]; ok && value != "true" {
```
Comment (janisz) [inline apps/app.go]: We should put warning that application will not be rgistered as 'true'

File: web/event_handler.go [hunk 3]
```diff
@@ -234,6 +191,61 @@ func (fh *EventHandler) handleStatusEvent(w http.ResponseWriter, body []byte) {
 	}
 }
 
+func (fh *EventHandler) handleDeploymentInfo(w http.ResponseWriter, body []byte) {
```
Comment (janisz) [inline web/event_handler.go]: Custom service name should be also handled here. Consider following situation. 
1. Deployed app `A` with label `consul=true`
2. Changed app label to `consul=B`

You end up with two services in consul: A and B.

File: web/event_handler.go [hunk 4]
```diff
@@ -234,6 +191,61 @@ func (fh *EventHandler) handleStatusEvent(w http.ResponseWriter, body []byte) {
...
+		fh.handleBadRequest(err, w)
+		return
+	}
+
+	log.Info("Got DeploymentInfoEvent")
+
+	errors := []error{}
+
+	for _, app := range deploymentInfo.StoppedApps() {
+		if app.IsConsulApp() {
```
Comment (janisz) [inline web/event_handler.go]: Body of this statement should be extracted to function

File: apps/app.go (hunk 2 above)
Comment (dankraw) [inline apps/app.go]: The behaviour is described in the README. How would you see the warning implemented - as an additional log?

Comment (janisz) [inline apps/app.go]: Yep, I was thinking about log and explicite mark this as deprecated and remove in next version

File: consul/consul.go (hunk 1 above)
Comment (janisz) [inline consul/consul.go]: OK

File: apps/app.go (hunk 2 above)
Comment (dankraw) [inline apps/app.go]: Added a warning to Consul.Register()

File: web/event_handler.go (hunk 3 above)
Comment (dankraw) [inline web/event_handler.go]: fixed by handling `deployment_step_success` for Restarted and renamed apps

Comment (dankraw) [inline web/event_handler.go]: thanks!

File: web/event_handler.go (hunk 4 above)
Comment (dankraw) [inline web/event_handler.go]: done

File: events/deployment_info.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+func (d *DeploymentEvent) appsMap(applications []*apps.App) map[apps.AppId]*apps.App {
+	result := make(map[apps.AppId]*apps.App)
+	for _, app := range applications {
+		result[app.ID] = app
+	}
+	return result
+}
+
+func (d *DeploymentEvent) consulAppsWithActionPerformed(allApps []*apps.App, actionType string) []*apps.App {
```
Comment (janisz) [inline events/deployment_info.go]: Nice

File: events/deployment_info.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+	if d.CurrentStep != nil {
+		if d.CurrentStep.Actions != nil {
+			return d.CurrentStep.Actions
+		}
+	}
+	return []*Action{}
+}
+
+func (d *DeploymentEvent) StoppedConsulApps() []*apps.App {
+	return d.consulAppsWithActionPerformed(d.OriginalApps(), "StopApplication")
```
Comment (janisz) [inline events/deployment_info.go]: Maybe we shoud extract actions to some const and link them with Marathon code https://github.com/mesosphere/marathon/blob/e3e867f00d4964d0367ec7e07519d6c9545681d6/src/main/scala/mesosphere/marathon/upgrade/DeploymentPlan.scala#L19-L34

**Comment by janisz** [comment]
LGTM but next time make smaller PR.
@adamdubiel what do you think?

File: events/deployment_info.go (hunk 6 above)
Comment (dankraw) [inline events/deployment_info.go]: I've linked our readme to the Marathon EventBus docs.

I was trying to implement these string literals as constants but it didn't make the code more readable at all (`events.STOP_APPLICATION_ACTION` for example). They appear only in this file (the API is with private scope) plus one more occurrence for both actions in a single test (for event creation purpose), so maybe let's just not do this. :) Or maybe you have an idea for a better implementation...

**Comment by dankraw** [comment]
@janisz This PR was meant to be more concise, sorry. ;)

**Comment by adamdubiel** [comment]
+1 on smaller PRs as a rule, but i don't feel entitled to give opinion about size of this particular one

**Comment by janisz** [comment]
@adamdubiel What do you think about code?

**Comment by dankraw** [comment]
@janisz 
IMHO @adamdubiel is focused on the feature being delivered, not reviewing these 1.5k lines of code. ;) I'm gonna merge it as it's already tested. We will hire someone else from Skylab for the upcoming issues to be resolved, so the reviewers team will grow. :)

**Pull request merged**